share/availability/test/withholding_adder.go

```diff
@@ -0,0 +1,79 @@
+package availability_test
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ipfs/boxo/blockservice"
+
+	"github.com/celestiaorg/celestia-app/pkg/da"
+	"github.com/celestiaorg/celestia-app/pkg/wrapper"
+	"github.com/celestiaorg/nmt"
+	"github.com/celestiaorg/rsmt2d"
+
+	"github.com/celestiaorg/celestia-node/libs/utils"
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/celestia-node/share/ipld"
+)
+
+// AddShares erasures and extends shares to blockservice.BlockService using the provided
+// ipld.NodeAdder.
+func AddAndWithholdShares(
+	ctx context.Context,
+	shares []share.Share,
+	adder blockservice.BlockService,
+	recoverable bool,
+) (*rsmt2d.ExtendedDataSquare, error) {
+	if len(shares) == 0 {
+		return nil, fmt.Errorf("empty data") // empty block is not an empty Data
+	}
+	squareSize := int(utils.SquareSize(len(shares)))
+	withheldSize := squareSize + 1
+	// create nmt adder wrapping batch adder with calculated size
+	batchAdder := ipld.NewNmtNodeAdder(ctx, adder, ipld.MaxSizeBatchOption(squareSize*2))
+	// create the nmt wrapper to generate row and col commitments
+	// recompute the eds
+	eds, err := rsmt2d.ComputeExtendedDataSquare(
+		shares,
+		share.DefaultRSMT2DCodec(),
+		wrapper.NewConstructor(uint64(squareSize),
+			nmt.NodeVisitor(batchAdder.Visit)),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failure to recompute the extended data square: %w", err)
+	}
+	// compute roots
+	_, err = eds.RowRoots()
+	if err != nil {
+		return nil, err
+	}
+	// commit the batch to ipfs
```

Comment:
It was only copied in the first case because the intention was to use some internal state, but the most recent version just iterates over the adder and ipld Translates the DAH. 

So I will see if I can just call AddShares and do the iteration after